components/tidb_query_datatype/src/codec/convert.rs

```diff
@@ -713,19 +709,23 @@ pub fn produce_str_with_specified_tp<'a>(
     // char count and truncate to flen chars if it is too long.
     if chs == charset::CHARSET_UTF8 || chs == charset::CHARSET_UTF8MB4 {
         let truncate_info = {
-            // In TiDB's version, the param `s` is a string,
-            // so we can unwrap directly here because we need the `s` represent a valid str
-            let s: &str = std::str::from_utf8(s.as_ref()).unwrap();
-            let mut indices = s.char_indices().skip(flen);
-            indices.next().map(|(truncate_pos, _)| {
-                let char_count = flen + 1 + indices.count();
-                (char_count, truncate_pos)
-            })
+            let s = &String::from_utf8_lossy(&s);
+            let mut truncate_pos = 0 as usize;
+            let mut char_indices = s.char_indices();
+            for _ in 0..flen {
+                let utf8_char = char_indices.next();
```

Comment:
could replace this `for loop` with `char_indices.take(flen).map(...)`

Reply:
Done